src/sinks/splunk_hec/common/service.rs

```diff
@@ -39,18 +88,59 @@ impl Service<HecRequest> for HecService {
     type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
 
     fn poll_ready(&mut self, _cx: &mut Context) -> std::task::Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
+        if self.max_pending_acks_reached.load(Ordering::Relaxed) {
+            Poll::Pending
+        } else {
+            Poll::Ready(Ok(()))
+        }
```

Comment:
So, this may currently work due to the number of wakeup sources present in `Driver`, but the code is at risk of entering a deadlocked state since the implicit contract of `poll_ready` is to ensure that when it returns `Poll::Pending`, something will wake-up the caller of `poll_ready` when there _is_ capacity.  As written, nothing is being told to wake up the caller when `self.max_pending_acks_reached` switches back to `false`.

A more robust strategy would be to use `tokio::sync::Semaphore`, wrapped with `tokio_util::sync::PollSemaphore`. A semaphore is essentially a fixed number of "permits" that can be acquired and returned.  If you try to acquire a permit and there are none available, you wait (and get woken up) when a permit becomes available.  In this case, you would construct the semaphore with a number of permits matching the "max pending acks" configuration value.

Thus, you would end up changing the service to try to acquire a permit in `poll_ready` via `PollSemaphore::poll_acquire`.  Since the permit needs to travel with the future generated by `call`, you'd have to store this internally in an `Option` or something, and then grab it in `call`.  Add the appropriate panics if `poll_ready` is called and there's already a permit stored, or if `call` is stored and there's no stored permit, etc.

Crucially, though, by calling `PollSemaphore::poll_acquire` in `poll_ready`, you would ensure that a wake-up chain was established so when there were no permits available, the tasdk would be woken up as soon as a permit was released back, when a future generated by `call` finally completed and was dropped.

Reply:
Thanks for the detailed explanation! I've refactored this in a3517c6. One question re:
> Add the appropriate panics if poll_ready is called and there's already a permit stored

The rustdoc describing `poll_ready` states that 

>  Once `poll_ready` returns `Poll::Ready(Ok(()))`, a request may be dispatched to the service using `call`. Until a request is dispatched, repeated calls to `poll_ready` must return either `Poll::Ready(Ok(()))` or `Poll::Ready(Err(_))`.

so, I decided to return `Poll::Ready(Err(_))` rather than panic if a permit is already stored. That said, would it be valid to instead check whether a permit is already stored and, if so, simply keep returning `Poll::Ready(Ok(()))`? Given the docs, it seems like users of a Service might expect calling `poll_ready` multiple times to be valid (although I guess I'm not sure why you'd continue doing so after the initial `Poll::Ready(Ok(()))`).

Reply:
Yeah, that's a fair point.  I've gone with panicking in the past because it usually indicates problems with the downstream implementation.

You can swap the `Err` for `Ok`, that's fine by me, so long as we don't throw away an already-acquired permit.